### ISSUE TITLE
Create `vmstatsd` app for collecting data in StatsD and DogStatsD formats

### ISSUE THREAD

**Issue by valyala** [VictoriaMetrics/VictoriaMetrics#206]
**Is your feature request related to a problem? Please describe.**
StatsD metrics ingestion format is in wide use now. It would be great if VictoriaMetrics could accept data in this format. There are third-party StatsD server implementations exist:
- The [original StatsD server written in node.js](https://github.com/statsd/statsd). This server can write data directly to VictoriaMetrics via [Graphite protocol](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/README.md#how-to-send-data-from-graphite-compatible-agents-such-as-statsd). However it has the following issues: `node.js`  and slowness.
- [statsd_exporter](https://github.com/prometheus/statsd_exporter). Prometheus can scrape data from it and then push the scraped data to VictoriaMetrics via `remote_write` API. The main issue with this implementation is operation complexity - you need to bring in Prometheus in the architecture.
- [DogStatsD](https://docs.datadoghq.com/developers/dogstatsd/). It supports popular extensions to StatsD such as [tags and new data types](https://docs.datadoghq.com/developers/dogstatsd/#dive-into-dogstatsd). But it is unclear whether it can push data to third-party storages such as VictoriaMetrics, or it can send data only to DataDog.

Such a diversity of possible solutions each with its own pros & cons complicates VictoriaMetrics usage for collecting StatsD metrics.

**Describe the solution you'd like**
It would be great to create a standalone officially supported `vmstatsd` app for collecting StatsD metrics and pushing them into VictoriaMetrics. The app must conform the following properties:
- It must be written in Go. This would simplify further maintenance.
- It must be optimized for speed and low memory usage.
- It must support input data types and formats from the original StatsD server.
- It must support DogStatsD extensions such as tags and additional data types (except of strings).
- It must be standalone, so it could be used with arbitrary time series databases that support Graphite plaintext protocol.

**Comment by valyala** [comment]
FYI, this is quite promising project - https://github.com/atlassian/gostatsd . It is written in go and it supports tags

**Comment by valyala** [comment]
Update: `vmstatsd` should be a part of [vmagent](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/app/vmagent/README.md) in the future.

**Comment by cristaloleg** [comment]
Another one, very lite and quite powerful https://github.com/smira/go-statsd

**Comment by n0madic** [comment]
@cristaloleg it is only a client, not a server ...

**Comment by faceair** [comment]
I don't think [statsd_exporter](https://github.com/prometheus/statsd_exporter) is very complicated, we can use vmagent scrape statsd_exporter. I think the main problem is that its histogram is too hard to use.

Another problem is that the original statsd does not support tagging, and the format for tagging is different in [different implementations](https://github.com/prometheus/statsd_exporter#tagging-extensions).  I prefer the datadog implementation because they wrote [the SDK in more languages](https://docs.datadoghq.com/developers/libraries/) !

Based on my conclusion above, I plan to implement statsd in vmagent like this:
1. support datadog's statsd format
2. the four metric types of count/gauge/timing/histogram are supported, with the histogram type using the VictoriaMetrics histogram.

We don't need to refer to any more statsd server implementations, just fork a copy of the [metrics](https://github.com/VictoriaMetrics/metrics) and change it.

I'm trying to implement it this way, feel free to discuss with me if you have more ideas! https://github.com/faceair/VictoriaMetrics/tree/vmstatsd

**Comment by faceair** [comment]
At times when statsd traffic is particularly high, we also need a pre-aggregation and routing component to help us forward metric with the same name to the same vmagent instance for final aggregation.
The good news is that it looks like most of the code for this component should be shared with the statsd part of the vmagent.

**Comment by valyala** [comment]
FYI, the next release of VictoriaMetrics and [vmagent](https://docs.victoriametrics.com/vmagent.html) will be able to accept metrics data from DataDog agent and DogStatsD. See [these docs](https://docs.victoriametrics.com/#how-to-send-data-from-datadog-agent).

**Comment by faceair** [comment]
https://github.com/VictoriaMetrics/VictoriaMetrics/commit/91b3c601bc1e2e0ea59052b7d94187e005fda894#diff-c0cb59743c2ebd0039da9e67acad6e09ccc92b0af9f0a3e39c2378c21b52d5cfR58

Good progress, but it doesn't look quite like what I want yet. I wish VictoriaMetrics would accept the original statsd protocol, so I wouldn't have to deploy [statsd](https://github.com/statsd/statsd) or [brubeck](https://github.com/github/brubeck) or [datadog-agent](https://github.com/DataDog/datadog-agent/tree/main/pkg/aggregator).

Does VictoriaMetrics still have plans to implement aggregator for statsd?

**Comment by kesavkolla** [comment]
Why is the restriction of single-node VictoriaMetrics? What if VictoriaMetrics running in cluster mode?

**Comment by lav-kush** [comment]
@valyala How do we configure endpoint of victoriametrics in Datadog agent submit metrics api in java?
Any reference would be appreciated.

**Comment by tenmozes** [comment]
@valyala should we close it as we introduced https://docs.victoriametrics.com/stream-aggregation.html

**Comment by valyala** [comment]
> @valyala should we close it as we introduced https://docs.victoriametrics.com/stream-aggregation.html

No, it would be great to add native support for statsd protocol in VictoriaMetrics. The https://github.com/VictoriaMetrics/VictoriaMetrics/pull/5053 looks promising in this regard.

**Comment by valyala** [comment]
FYI, the [pull request](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5052) has been reverted in the commit 2da7dfc7548cf8dda523d012e39eb7059289f933 because of the following reasons:

There are many statsd servers exist:

- https://github.com/statsd/statsd - classical statsd server
- https://docs.datadoghq.com/developers/dogstatsd/ - statsd server from DataDog built into DatDog Agent ( https://docs.datadoghq.com/agent/ )
- https://github.com/avito-tech/bioyino - high-performance statsd server
- https://github.com/atlassian/gostatsd - statsd server in Go
- https://github.com/prometheus/statsd_exporter - statsd server, which exposes the aggregated data as Prometheus metrics

These servers can be used for efficient aggregating of statsd data and sending it to VictoriaMetrics according to https://docs.victoriametrics.com/#how-to-send-data-from-graphite-compatible-agents-such-as-statsd (the https://github.com/prometheus/statsd_exporter can be scraped as usual Prometheus target according to https://docs.victoriametrics.com/#how-to-scrape-prometheus-exporters-such-as-node-exporter ).

Adding support for statsd data ingestion protocol into VictoriaMetrics makes sense only if it provides significant advantages over the existing statsd servers, while has no significant drawbacks comparing to existing statsd servers.

The main advantage of statsd server built into VictoriaMetrics and vmagent - getting rid of additional statsd server.

The main drawback is non-trivial and inconvenient streaming aggregation configs, which must be used for the ingested statsd metrics (see https://docs.victoriametrics.com/stream-aggregation/ ). These configs are incompatible with the configs for standalone statsd servers, so you need to manually translate configs of the used statsd server to stream aggregation configs when migrating from standalone statsd server to statsd server built into VictoriaMetrics (or [vmagent](https://docs.victoriametrics.com/vmagent/)).

Another important drawback is that it is very easy to shoot yourself in the foot when using built-in statsd server with the `-statsd.disableAggregationEnforcement` command-line flag or with improperly configured streaming aggregation. In this case the ingested statsd metrics will be stored to VictoriaMetrics as is without any aggregation. This may result in high CPU usage during data ingestion, high disk space usage for storing all the unaggregated statsd metrics and high CPU usage during querying, since all the unaggregated metrics must be read, unpacked and processed during querying.

P.S. Built-in statsd server can be added to VictoriaMetrics and vmagent after figuring out more ergonomic specialized configuration for aggregating of statsd metrics. The main requirements for this configuration:

- easy to write, read and update (ideally it should work out of the box for most cases without additional configuration)
- hard to misconfigure (e.g. hard to shoot yourself in the foot)

It would be great if this configuration will be compatible with the configuration of the most widely used statsd server.

In the mean time it is recommended continue using external statsd server, which sends aggregated metrics to VictoriaMetrics or vmagent according to [these docs](https://docs.victoriametrics.com/#how-to-send-data-from-graphite-compatible-agents-such-as-statsd) (use [these docs](https://docs.victoriametrics.com/#how-to-send-data-from-datadog-agent) for sending dogstatsd metrics from DataDog agent).
